### PR TITLE
fix(schedules): generate return trips when path times are fractional

### LINKED ISSUE #2005: Impossible to generate symetric schedules
When we select symetric schedules, it only creates schedules in one direciton, event when both outbound and inbound paths are provided.

Also, for some reason, it is now impossible to gnereate any schedules for "whole day/journée complète" periods.

### PR DESCRIPTION
**Pull request by kaligrafy**
Symmetric schedule generation compares unit.timeInCycle (integer) to operatingTimeWithLayoverTimeSeconds, which can be fractional after path geography recalculation. Ceil both outbound and inbound cycle times to whole seconds before scheduling so return trips are generated.

Closes #2005

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved consistency when rounding travel times, distances, schedules, transfers, and accessibility calculations.
  * Fractional schedule durations now produce reliable trip timings and vehicle requirements.
  * Invalid or unavailable routing results are handled safely instead of generating incorrect values.
  * Deadhead travel times with missing or negative durations are recorded as unavailable.
  * Path and routing calculations now better handle fractional and non-finite values.

* **Tests**
  * Added coverage for fractional schedules, routing edge cases, deadhead travel times, and invalid numeric inputs.

### CODE REVIEW

**Review by tahini — COMMENTED** [review]
Ça fonctionne, mais ne faudrait-il pas plutôt que le temps à la base ne soit pas fractionnel?

**Comment by kaligrafy** [comment]
> Ça fonctionne, mais ne faudrait-il pas plutôt que le temps à la base ne soit pas fractionnel?
Oui, j'ai corrigé partout, les distances et les temps.

File: packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts [hunk 1]
```diff
@@ -473,6 +473,37 @@ describe('generateForPeriod', () => {
...
+        testAttributes.periods = testPeriods;
+
+        const schedule = new Schedule(testAttributes, true, localCollectionManager);
+        const { trips } = schedule.generateForPeriod(testPeriods[0].period_shortname as string);
+
+        const outboundTrips = trips.filter((trip) => trip.path_id === outbound.getId());
+        const inboundTrips = trips.filter((trip) => trip.path_id === inbound.getId());
+        expect(outboundTrips.length).toBeGreaterThan(0);
+        expect(inboundTrips.length).toBeGreaterThan(0);
+    });
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts]: verify if fixed

File: packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts [hunk 2]
```diff
@@ -473,6 +473,37 @@ describe('generateForPeriod', () => {
...
+        testPeriods[0].inbound_path_id = inbound.getId();
+        testAttributes.periods = testPeriods;
+
+        const schedule = new Schedule(testAttributes, true, localCollectionManager);
+        const { trips } = schedule.generateForPeriod(testPeriods[0].period_shortname as string);
+
+        const outboundTrips = trips.filter((trip) => trip.path_id === outbound.getId());
+        const inboundTrips = trips.filter((trip) => trip.path_id === inbound.getId());
+        expect(outboundTrips.length).toBeGreaterThan(0);
+        expect(inboundTrips.length).toBeGreaterThan(0);
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts]: verify if fixed

File: packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts]: fixed

File: packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-common/src/services/schedules/__tests__/Schedule.test.ts]: fixed

File: packages/chaire-lib-common/src/utils/__tests__/DateTimeUtils.test.ts [hunk 3]
```diff
@@ -7,22 +7,22 @@
...
-  expect(DateTimeUtils.decimalHourToTimeStr(23.999)).toBe('23:59');
-  expect(DateTimeUtils.decimalHourToTimeStr(6.999, true, true)).toBe('6:59:56');
-  expect(DateTimeUtils.decimalHourToTimeStr(6.9999, true, true)).toBe('7:00:00');
+test('should convert decimal hour to time string', () => {
+    expect(DateTimeUtils.decimalHourToTimeStr(6.23)).toBe('6:13');
+    expect(DateTimeUtils.decimalHourToTimeStr(6.999)).toBe('6:59');
+    expect(DateTimeUtils.decimalHourToTimeStr(23.99)).toBe('23:59');
+    expect(DateTimeUtils.decimalHourToTimeStr(23.999)).toBe('23:59');
+    expect(DateTimeUtils.decimalHourToTimeStr(6.999, true, true)).toBe('6:59:56');
+    expect(DateTimeUtils.decimalHourToTimeStr(6.9999, true, true)).toBe('7:00:00');
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/chaire-lib-common/src/utils/__tests__/DateTimeUtils.test.ts:16]: hors scope de la PR.

File: packages/transition-common/src/services/line/Line.ts [hunk 4]
```diff
@@ -228,7 +229,7 @@ export class Line extends ObjectWithHistory<LineAttributes> implements Saveable
                         }
 
                         deadHeadTravelTimesBetweenPathsByPathId[path1Id][path2Id] =
-                            travelTimeSeconds !== null ? Math.ceil(travelTimeSeconds) : null;
+                            travelTimeSeconds !== null ? ceilToPositiveInteger(travelTimeSeconds) : null;
```
[the remark this answers is not in the document]
Comment (kaligrafy) [reply packages/transition-common/src/services/line/Line.ts:232]: fixed

File: packages/transition-backend/src/services/path/PathGtfsGeographyGenerator.ts [hunk 5]
```diff
@@ -215,10 +216,12 @@ const buildPathData = (params: {
         dwellTimeSeconds: dwellTimeSecondsData,
         layoverTimeSeconds: layoverTimeSeconds,
         travelTimeWithoutDwellTimesSeconds: totalTravelTimeWithoutDwellTimesSeconds,
-        totalDistanceMeters: Math.ceil(totalDistanceMeters),
+        totalDistanceMeters: ceilToPositiveInteger(totalDistanceMeters),
         totalDwellTimeSeconds: totalDwellTimeSeconds,
-        operatingTimeWithoutLayoverTimeSeconds: totalTravelTimeWithDwellTimesSeconds,
-        operatingTimeWithLayoverTimeSeconds: totalTravelTimeWithDwellTimesSeconds + layoverTimeSeconds,
+        operatingTimeWithoutLayoverTimeSeconds: ceilToPositiveInteger(totalTravelTimeWithDwellTimesSeconds),
```
Comment (tahini) [inline packages/transition-backend/src/services/path/PathGtfsGeographyGenerator.ts]: Le totalTravelTimeWithDwellTimeSeconds et layoverTimeSeconds sont probablement la somme de plusieurs valeurs. En faisant un ceil ici, la somme des valeurs va demeurer potentiellement fractionnelle et donc le total ne matchera plus la somme. Ne faudrait-il pas plutôt ceiler les valeurs qui composent cette somme (genre durées de segments)? Puisque c'est en secondes, on n'a pas besoind e fraction de secondes, nous ne sommes pas encore à ce genre d'échelle en transport de personnes

File: packages/transition-common/src/services/path/PathGeographyGenerator.ts [hunk 6]
```diff
@@ -523,10 +529,10 @@ const buildSegmentsAndGeometry = (
         appendLegCoordinates(leg, globalCoordinates);
 
         segmentTimeAndDistance.travelTimeSeconds += leg.duration;
-        segmentTimeAndDistance.distanceMeters = (segmentTimeAndDistance.distanceMeters || 0) + Math.ceil(leg.distance);
-
+        segmentTimeAndDistance.distanceMeters = (segmentTimeAndDistance.distanceMeters || 0) + leg.distance;
```
Comment (tahini) [inline packages/transition-common/src/services/path/PathGeographyGenerator.ts:530]: C'est peut-être la ligne en haut `+= leg.duration` qu'il faudrait ceiler, en mettant un entier dans le travelTimeSeconds

**Review by tahini — CHANGES_REQUESTED** [review]

File: packages/transition-backend/src/services/path/PathGtfsGeographyGenerator.ts (hunk 5 above)
Comment (kaligrafy) [reply packages/transition-backend/src/services/path/PathGtfsGeographyGenerator.ts]: corrigé

File: packages/transition-common/src/services/path/PathGeographyGenerator.ts (hunk 6 above)
Comment (kaligrafy) [reply packages/transition-common/src/services/path/PathGeographyGenerator.ts:530]: corrigé

File: packages/chaire-lib-common/src/utils/__tests__/MathUtils.test.ts [hunk 7]
```diff
@@ -123,4 +123,26 @@ test('should get sequential array', function() {
...
+    expect(MathUtils.ceilToPositiveInteger(input)).toBe(expected);
+});
+
+test.each([
+    [-5],
+    [-0.1],
+    [NaN],
+    [Infinity],
+    [-Infinity]
+])('should throw for invalid input %p', (input) => {
```
Comment (tahini) [inline packages/chaire-lib-common/src/utils/__tests__/MathUtils.test.ts:146]: Ce test n'étant pas dans un describe, il manque l'information sur la fonction testée qui est ici ceilToPostivieInteger

File: packages/chaire-lib-common/src/utils/DateTimeUtils.ts [hunk 8]
```diff
@@ -92,7 +93,7 @@ const secondsToHoursDecimal = function (seconds) {
     return _isFinite(seconds) ? seconds / 3600 : null;
 };
 
-const minutesToHours = function (minutes: string | number, rounding = Math.ceil) {
+const minutesToHours = function (minutes: string | number, rounding = ceilToPositiveInteger) {
```
Comment (tahini) [inline packages/chaire-lib-common/src/utils/DateTimeUtils.ts:96]: Ici, on change un peu le contrat de cette fonction. Je ne sais pas si tous les appelants sont toujours avec des valeurs positives, mais advenant qqn qui utilisait la version sans passer de fonction, mais avec une valeur négative, et faisait un check éventuel de cette valeur dans son code path (ou pas si on veut des temps négatifs), il y aurait maintenant une exception. As-tu vérifié les appels, dans Transition ET Evolution pour être sûr que ça n'impacte pas? Sinon, c'est un breaking change qu'il faudrait gérer... Ça se peut que ce soit correct, mais au moins dans le repo Transition s'assurer que tout est beau, et ouvrir un issue dans Evolution pour la prochaine mise à jour de Transition.

File: packages/transition-common/src/services/line/Line.ts [hunk 9]
```diff
@@ -47,7 +48,7 @@ export interface LineAttributes extends GenericAttributes {
     is_enabled?: boolean;
     data: {
         gtfs?: GtfsRoute;
-        deadHeadTravelTimesBetweenPathsByPathId?: { [key: string]: { [key: string]: number } };
+        deadHeadTravelTimesBetweenPathsByPathId?: { [key: string]: { [key: string]: number | null } };
```
Comment (tahini) [inline packages/transition-common/src/services/line/Line.ts:51]: Le moment serait bien choisi d'ajouter de la documentation sur ce champ, qu'est-ce que ça mange en hiver? pour qqn qui connaît pas les termes de transport collectif et ne veut pas lire le latex en entier...

File: packages/transition-common/src/services/line/Line.ts [hunk 10]
```diff
@@ -219,16 +220,16 @@ export class Line extends ObjectWithHistory<LineAttributes> implements Saveable
                             overview: 'full'
                         });
 
-                        const travelTimeSeconds = _get(routingResult, 'routes[0].duration', null);
-
-                        if (travelTimeSeconds === null) {
+                        let travelTimeSeconds = _get(routingResult, 'routes[0].duration', null);
+                        if (travelTimeSeconds === -1 || travelTimeSeconds === null) {
```
Comment (tahini) [inline packages/transition-common/src/services/line/Line.ts:224]: Peut-être < 0 au lieu de === -1?

File: packages/transition-common/src/services/path/PathGeographyGenerator.ts [hunk 11]
```diff
@@ -373,6 +375,7 @@ const buildPathData = (
         travelTimeWithoutDwellTimesSeconds: totals.totalTravelTimeWithoutDwellTimesSeconds,
         totalDistanceMeters: totals.totalDistance,
         totalDwellTimeSeconds: totals.totalDwellTimeSeconds,
+        // Totals are sums of already-integer segment times, dwells and layover.
```
Comment (tahini) [inline packages/transition-common/src/services/path/PathGeographyGenerator.ts:378]: Commentaire d'IA non-nécessaire ici

File: packages/transition-common/src/services/path/PathGeographyGenerator.ts [hunk 12]
```diff
@@ -522,11 +525,13 @@ const buildSegmentsAndGeometry = (
 
         appendLegCoordinates(leg, globalCoordinates);
 
+        // Routed duration stays fractional: it is the speed input for accel/decel physics.
```
Comment (tahini) [inline packages/transition-common/src/services/path/PathGeographyGenerator.ts:528]: tu dis ça, mais 3 lignes plus loin, dans le `if` ces mêmes valeurs sont transformées en entier. Dans un monde idéal, si le type en question (le type de `segmentTimeAndDistance`) est utilisé avec des entiers, il faudrait les valeurs soient des entiers dès le départ, sinon, on ne peut pas faire d'assomptions.

File: packages/transition-common/src/services/path/PathGeographyGenerator.ts [hunk 13]
```diff
@@ -619,6 +625,10 @@ const adjustSegmentTime = (
         current.segmentsData[segmentIndex].travelTimeSeconds =
             current.segmentsData[segmentIndex].travelTimeSeconds * current.ratioDifferenceTime;
     }
+    // Store integer seconds so path totals equal the sum of their parts.
+    current.segmentsData[segmentIndex].travelTimeSeconds = ceilToPositiveInteger(
```
Comment (tahini) [inline packages/transition-common/src/services/path/PathGeographyGenerator.ts:629]: Ici aussi, il y a un `if`/`else` juste en haut qui termine en assignant le `travelTimeSeconds`, juste avant de faire un ceil dessus. Ce serait mieux de mettre le `ceil` à la source. D'ailleurs, je crois que ça s'applique juste au `else`, le `if` devrait déjà être ok? à moins que adjustment puisse être fractionnel.

**Review by tahini — CHANGES_REQUESTED** [review]

File: packages/transition-common/src/services/path/PathGeographyGenerator.ts [hunk 14]
```diff
@@ -555,12 +559,12 @@ const buildSegmentsAndGeometry = (
         segments.push(segmentCoordinatesStartIndex);
         segmentsData.push({
             travelTimeSeconds: result.duration.calculatedSegmentDurationSeconds,
-            distanceMeters: segmentTimeAndDistance.distanceMeters
```
Comment (tahini) [inline packages/transition-common/src/services/path/PathGeographyGenerator.ts:558]: Est-ce que le `travelTimeSeconds` plus haut a besoin d'être ceiler aussi? Ou il est correct? Le `segmentDistanceMeters`, lui a été ceiler à la ligne 542, mais le temps est le résultat d'un calcul que je ne suis pas certaine s'il retourne un entier. À valider svp

**Review by tahini — APPROVED** [review]
C'est bon, juste un endroit à valider svp, pour savoir si c'est fractionnel ou pas. Sinon, tu as testé et ça fonctionne pour les cédules?

File: packages/transition-common/src/services/path/PathGeographyGenerator.ts (hunk 14 above)
Comment (kaligrafy) [reply packages/transition-common/src/services/path/PathGeographyGenerator.ts:558]: corrigé!
